rollout/temlateref.go

```diff
@@ -198,36 +209,51 @@ func (r *informerBasedTemplateResolver) newInformerForGVK(gvk schema.GroupVersio
 		nil)
 	informer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
-			r.requeueReferencedRollouts(obj, gvk)
+			r.updateRolloutsReferenceAnnotation(obj, gvk)
 		},
 		UpdateFunc: func(oldObj, newObj interface{}) {
-			r.requeueReferencedRollouts(newObj, gvk)
+			r.updateRolloutsReferenceAnnotation(newObj, gvk)
 		},
 		DeleteFunc: func(obj interface{}) {
-			r.requeueReferencedRollouts(obj, gvk)
+			r.updateRolloutsReferenceAnnotation(obj, gvk)
 		},
 	})
 	return informer, nil
 
 }
 
-// requeueReferencedRollouts re-queues all rollouts referenced by given object
-func (r *informerBasedTemplateResolver) requeueReferencedRollouts(obj interface{}, gvk schema.GroupVersionKind) {
-	roMeta, err := meta.Accessor(obj)
+// updateRolloutsReferenceAnnotation update the annotation of all rollouts referenced by given object
+func (r *informerBasedTemplateResolver) updateRolloutsReferenceAnnotation(obj interface{}, gvk schema.GroupVersionKind) {
+	workloadMeta, err := meta.Accessor(obj)
 	if err != nil {
 		return
 	}
+
 	rollouts, err := r.rolloutsInformer.GetIndexer().ByIndex(templateRefIndexName, refKey(v1alpha1.ObjectRef{
 		Kind:       gvk.Kind,
 		APIVersion: gvk.GroupVersion().String(),
-		Name:       roMeta.GetName(),
-	}, roMeta.GetNamespace()))
+		Name:       workloadMeta.GetName(),
+	}, workloadMeta.GetNamespace()))
 	if err != nil {
 		return
 	}
+
+	generation := strconv.FormatInt(workloadMeta.GetGeneration(), 10)
 	for _, ro := range rollouts {
-		if key, err := cache.MetaNamespaceKeyFunc(ro); err == nil {
-			r.rolloutWorkQueue.Add(key)
+		un, ok := ro.(*unstructured.Unstructured)
```

Comment:
@jessesuen , in the e2e test, the ro can be mappted to `*unstructured.Unstructured` without any issue. However, in the test code, the rollout can't be resolved. Do you have any idea on this? Thanks.

Reply:
FYI, to reproduce: `go test -timeout 999s -run ^TestRequeueReferencedRollouts$ ./rollout -v`